### PR TITLE
SPARQL -> FQL translator refactor

### PR DESCRIPTION
**Pull request by dpetran**
This is a rewrite of the SPARQL translator that solves several issues, and should make adding support for more SPARQL syntax easier.

Specifically fixes :union patterns in where clauses, subject iri parsing, and REGEX functions.

fixes https://github.com/fluree/core/issues/90

### CODE REVIEW

File: src/clj/fluree/db/query/fql/syntax.cljc [hunk 1]
```diff
@@ -131,12 +131,12 @@
...
                        [:selector ::selector]
                        [:collection [:sequential ::selector]]]
-    ::direction       [:orn {:error/message "Direction must be ASC or DESC"}
+    ::direction       [:orn {:error/message "Direction must be \"asc\" or \"desc\""}
                        [:asc [:fn asc?]]
                        [:desc [:fn desc?]]]
     ::ordering        [:orn {:error/message "Ordering must be a var or two-tuple formatted ['ASC' or 'DESC', var]"}
                        [:scalar ::var]
-                       [:vector [:and list?
+                       [:vector [:and sequential?
```
Comment (zonotope) [inline src/clj/fluree/db/query/fql/syntax.cljc:139]: I think the ordering must be an explicit list to differentiate it from our syntactic usage of vectors for things like groups of variables.

File: test/fluree/db/query/json_ld_compound_test.clj [hunk 2]
```diff
@@ -118,7 +118,7 @@
                                  :select  [?name ?favNums]
                                  :where   {:schema/name ?name
                                            :ex/favNums  ?favNums}
-                                 :orderBy (desc ?favNums)})
+                                 :orderBy [(desc ?favNums)]})
```
Comment (zonotope) [inline test/fluree/db/query/json_ld_compound_test.clj:121]: Does this change require ordering clauses to be wrapped in a vector now?

Comment (zonotope) [reply test/fluree/db/query/json_ld_compound_test.clj:121]: Did you change it? If so, could you change this test back too?

Comment (cap10morgan) [reply test/fluree/db/query/json_ld_compound_test.clj:121]: This still needs to be done, it looks like?

File: test/fluree/db/query/sparql_test.cljc [hunk 3]
```diff
@@ -140,6 +140,20 @@
...
+                 FROM <cookbook/base>
+                 WHERE {
+                   ?s <@type> ?t.
+                   ?s schema:name ?name.
+                   FILTER REGEX(?name, \"^Jon\", \"i\")
+                 }"
+          {:keys [where]} (sparql/->fql query)]
+      (is (= [{"@id" "?s", "@type" "?t"}
+              {"@id" "?s", "schema:name" "?name"}
+              [:filter ["(regex ?name \"^Jon\" \"i\")"]]]
```
Comment (cap10morgan) [inline test/fluree/db/query/sparql_test.cljc:156]: What is this testing? Just that REGEX filters are translated correctly? Might be good to add a `testing` annotation (or similar) to that effect. Looks like the `psm` prefix is unused, is that intentional?

File: test/fluree/db/query/sparql_test.cljc [hunk 4]
```diff
@@ -233,14 +247,14 @@
...
-        (is (= ["asc" "?favNums"]
+        (is (= [["asc" "?favNums"]]
                orderBy))))
     (testing "DESC"
       (let [query "SELECT ?favNums
                    WHERE {?person person:favNums ?favNums}
                    ORDER BY DESC(?favNums)"
             {:keys [orderBy]} (sparql/->fql query)]
-        (is (= ["desc" "?favNums"]
+        (is (= [["desc" "?favNums"]]
```
Comment (cap10morgan) [inline test/fluree/db/query/sparql_test.cljc:257]: Looks like these need to be changed back to single-dimensional vectors too?

File: test/fluree/db/query/sparql_test.cljc [hunk 5]
```diff
@@ -254,7 +268,7 @@
                  WHERE {?e person:favNums ?favNums.}
                  GROUP BY ?e HAVING(SUM(?favNums) > 1000)"
           {:keys [groupBy having]} (sparql/->fql query)]
-      (is (= "?e"
+      (is (= ["?e"]
```
Comment (cap10morgan) [inline test/fluree/db/query/sparql_test.cljc:271]: Needs to be reverted too?

File: test/fluree/db/query/sparql_test.cljc [hunk 6]
```diff
@@ -296,6 +311,24 @@
...
 
+(deftest subject-iri
+  (let [query
+        "PREFIX psm: <http://srv.ktbl.de/data/psm/>
+PREFIX ex: <http://example.org/>
+
+SELECT ?p ?o
+FROM <cookbook/base>
+WHERE {
+    ex:andrew ?p ?o. # Should provide just the triples related to andrew
```
Comment (cap10morgan) [inline test/fluree/db/query/sparql_test.cljc:322]: This comment might be more useful in a `testing` annotation?

**Review by cap10morgan — COMMENTED** [review]
Looks like there are still some outstanding TODOs in here

File: test/fluree/db/query/sparql_test.cljc (hunk 5 above)
Comment (dpetran) [reply test/fluree/db/query/sparql_test.cljc:271]: Same deal here, the translator will always output a vector for ORDER BY, even though FQL accepts it both ways.

File: test/fluree/db/query/sparql_test.cljc (hunk 6 above)
Comment (dpetran) [reply test/fluree/db/query/sparql_test.cljc:322]: This was a direct test case from the ticket, I think I'll add it in with the `simple-triple` expectations.

File: src/clj/fluree/db/query/sparql/translator.cljc [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+
+(defn literal-quote
+  "Quote a non-variable string literal for use in an expression."
+  [x]
+  (if (and (string? x)
+           (not (str/starts-with? x "?")))
+    (str "\"" x "\"")
+    x))
+
+;; take a term rule and return a value
```
Comment (zonotope) [inline src/clj/fluree/db/query/sparql/translator.cljc:19]: This comment should go into the docstring of the `parse-term` multimethod.

File: src/clj/fluree/db/query/sparql/translator.cljc [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+(defmulti parse-term (fn [[tag & _]] tag))
+;; take a rule return a sequence of entry tuples [[k v]...]
```
Comment (zonotope) [inline src/clj/fluree/db/query/sparql/translator.cljc:21]: This comment should go into the docstring of the `parse-rule` multimethod

**Review by zonotope — APPROVED** [review]

**Comment by dpetran** [comment]
The CI was stuck waiting for a container so I merged and ran CI locally. All commits are now on main.

**Pull request closed without merging**